### PR TITLE
Log the correct learning rate

### LINKED ISSUE #35942: Learning rate logging off by one training step
### System Info

transformers==4.48.1

### Who can help?

@muellerz @SunMarc

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The Trainer class steps the learning rate scheduler at the end of each training loop iteration before calling `Trainer._maybe_log_save_evaluate()`. As a result, the logged learning rate is the learning rate of the next training step, not the one that was just performed. This off-by-one error is usually not much of a problem with monotonic schedulers, but with cycling schedulers such as `cosine_with_restarts`, it can be the difference between something very close to zero and the max learning rate at cycle boundaries.

LR scheduler is stepped here: https://github.com/huggingface/transformers/blob/4ec425ffad56cdbedfb97ab2d11243e42889f71c/src/transformers/trainer.py#L2605

Logging and checkpointing occurs here: https://github.com/huggingface/transformers/blob/4ec425ffad56cdbedfb97ab2d11243e42889f71c/src/transformers/trainer.py#L2611

### Expected behavior

The call to `Trainer._maybe_log_save_evaluate()` should be moved above the LR scheduler step (or vice versa), otherwise an incorrect learning rate will be logged.

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?

> The Trainer class steps the learning rate scheduler at the end of each training loop iteration before calling Trainer._maybe_log_save_evaluate(). As a result, the logged learning rate is the learning rate of the next training step, not the one that was just performed. This off-by-one error is usually not much of a problem with monotonic schedulers, but with cycling schedulers such as cosine_with_restarts, it can be the difference between something very close to zero and the max learning rate at cycle boundaries.

Fixes https://github.com/huggingface/transformers/issues/35942
Similar to https://github.com/huggingface/transformers/pull/30169

### CODE REVIEW

**Review by muellerzr — APPROVED** [review]
Looks good on my end, thanks for giving this a go!

**Comment by SunMarc** [comment]
This might potentially break trl dpo_online trainer I think @qgallouedec

**Comment by qgallouedec** [comment]
Thank you! I will check and correct TRL if necessary. But feel free to merge, I can do it once merged.
